packages/core/src/types.ts

```diff
@@ -75,6 +73,381 @@ export interface PositionWithOffset extends Position {
+ * - `suggestion` means the rule suggests an alternate or better approach.
+ * - `layout` means the rule is looking at spacing, indentation, etc.
+ */
+export type RuleType = "problem" | "suggestion" | "layout";
+
+/**
+ * The type of fix the rule can provide:
+ * - `code` means the rule can fix syntax.
+ * - `whitespace` means the rule can fix spacing and indentation.
+ */
+export type RuleFixType = "code" | "whitespace";
+
+/* eslint-disable @typescript-eslint/consistent-indexed-object-style -- Needs to be interface so people can extend it. */
+/**
+ * An object containing visitor information for a rule. Each method is either the
+ * name of a node type or a selector, or is a method that will be called at specific
+ * times during the traversal.
+ */
+export interface RuleVisitor {
+	/**
+	 * Called for each node in the AST or at specific times during the traversal.
+	 */
+	[key: string]:
+		| ((node: unknown, parent?: unknown) => void)
+		| ((...unknown: unknown[]) => void);
+}
+/* eslint-enable @typescript-eslint/consistent-indexed-object-style -- Needs to be interface so people can extend it. */
+
+/**
+ * Rule meta information used for documentation.
+ */
+export interface RulesMetaDocs {
+	/**
+	 * A short description of the rule.
+	 */
+	description?: string | undefined;
+
+	/**
+	 * The URL to the documentation for the rule.
+	 */
+	url?: string | undefined;
+
+	/**
+	 * The category the rule falls under.
+	 * @deprecated No longer used.
+	 */
+	category?: string | undefined;
+
+	/**
+	 * Indicates if the rule is generally recommended for all users.
+	 */
+	recommended?: boolean | undefined;
+}
+
+/**
+ * Meta information about a rule.
+ */
+export interface RulesMeta {
+	/**
+	 * Properties that are used when documenting the rule.
+	 */
+	docs?: RulesMetaDocs | undefined;
+
+	/**
+	 * The type of rule.
+	 */
+	type?: RuleType | undefined;
+
+	/**
+	 * The schema for the rule options. Required if the rule has options.
+	 */
+	schema?: JSONSchema4 | JSONSchema4[] | false | undefined;
+
+	/**
+	 * The messages that the rule can report.
+	 */
+	messages: Record<string, string>;
+
+	/**
+	 * The deprecated rules for the rule.
+	 */
+	deprecated?: boolean | undefined;
+
+	/**
+	 * When a rule is deprecated, indicates the rule ID(s) that should be used instead.
+	 */
+	replacedBy?: string[] | undefined;
+
+	/**
+	 * Indicates if the rule is fixable, and if so, what type of fix it provides.
+	 */
+	fixable?: RuleFixType | undefined;
+
+	/**
+	 * Indicates if the rule may provide suggestions.
+	 */
+	hasSuggestions?: boolean | undefined;
+}
+
+/**
+ * Represents the context object that is passed to a rule. This object contains
+ * information about the current state of the linting process and is the rule's
+ * view into the outside world.
+ */
+export interface RuleContext<
+	LangOptions = LanguageOptions,
+	Code extends SourceCode = SourceCode,
+	RuleOptions = unknown[],
+	Node = unknown,
+> {
```

Comment:
> `LangOptions, Code, RuleOptions, Node`

👍 confirming that, as unweildy as having four type parameters feels, I think this is the right way to represent things. Each rule definition is specific to the _language_, _source code type_, _rule options_, and _language node_.

I'll note that a somewhat-equivalent-ish syntax to represent passing many type parameters would be to pass a single type parameter extending a shape. Roughly:

```suggestion
export interface RuleContextSettings {
	LangOptions: LanguageOptions;
	Code?: SourceCode;
	RuleOptions: unknown[];
	Node?: unknown;
}

/**
 * Represents the context object that is passed to a rule. This object contains
 * information about the current state of the linting process and is the rule's
 * view into the outside world.
 */
export interface RuleContext<Settings extends RuleContextSettings> {
```

...then later on referring to members of that type parameter:

```diff
- sourceCode: Code;
+ sourceCode: Settings["Code"];
```

It's conceptually like the refactor from a function with four parameter to a function with a single object parameter:

```diff
- function ruleContext(langOptions, code, ruleOptions, node) {
+ function ruleContext(settings) {

- langOptions;
+ settings.langOptions;
```

I've put up a comparison of what that could look like here: https://github.com/eslint/rewrite/compare/types-update-2...JoshuaKGoldberg:rewrite:types-update-settings-type-parameters.

Not saying either way is better, just showing another syntax as reference in case you prefer it.

Reply:
Ah, interesting! I think I prefer the way it is, currently, but I consider myself educated. Thanks!

Reply:
Actually, would that mean the type parameters could be specified in any order?

Reply:
Yeah! And you're forced to use explicit names for them. `RuleContext<{ Key: Type, ... }>` instead of `RuleContext<Type, ...>`.

Reply:
Ooh nice! Okay, going to change to that.

Reply:
I've switched all the types with way-too-may generic parameters to this model. The downside (as far as I can tell), is that TypeScript then either requires all of the properties or none of them -- the default is only used if nothing is passed, so I'm unable to assign defaults for each property. :thinking: